src/main/java/io/tiledb/spark/TileDBDataWriter.java

```diff
@@ -456,25 +459,43 @@ public void write(InternalRow record) throws IOException {
 
   private void flushBuffers() throws TileDBError {
     this.metricsUpdater.startTimer(queryWriteFlushBuffersTimerName);
+    JavaArray jArray;
     long buffersInBytes = 0;
-    query.setBuffer(Constants.TILEDB_COORDS, nativeArrayBuffers[0], nRecordsBuffered * nDims);
+
+    query.setBuffer(
+        Constants.TILEDB_COORDS,
+        new NativeArray(
+            ctx,
+            javaArrayBuffers[0].get(),
+            javaArrayBuffers[0].getDataType(),
+            javaArrayBuffers[0].getNumElements()),
+        nRecordsBuffered * nDims);
     // Calculate bytes we are writing for metrics starting with dimension
     buffersInBytes += nRecordsBuffered * nDims * bufferDatatypes[nDims - 1].getNativeSize();
     for (int i = nDims; i < bufferNames.length; i++) {
       String name = bufferNames[i];
       // Calculate bytes we are writing for metrics
       buffersInBytes += nRecordsBuffered * nDims * bufferDatatypes[i].getNativeSize();
 
+      Datatype bufferDataType = javaArrayBuffers[i].getDataType();
+      Object bufferData =
+          javaArrayBuffers[i].getDataType() == Datatype.TILEDB_CHAR
+              ? new String((byte[]) javaArrayBuffers[i].get())
+              : javaArrayBuffers[i].get();
+
       boolean isVar = (bufferValNum[i] == Constants.TILEDB_VAR_NUM);
       if (isVar) {
         query.setBuffer(
             name,
-            nativeArrayOffsetBuffers[i],
-            nativeArrayBuffers[i],
+            new NativeArray(ctx, javaArrayOffsetBuffers[i], Datatype.TILEDB_UINT64),
+            new NativeArray(ctx, bufferData, bufferDataType, javaArrayBuffers[i].getNumElements()),
```

Comment:
is `getNumElements` here correct? The results probably aren't the full size of the data, shouldn't this be `nativeArrayBufferElements[i]`?